REGRESSION (276531@main): Content flash when expanding image viewer on eBay page

graouts opened the pull request:
#### d08faf9fab62fcbe230979774bacbdc53684e816
<pre>
REGRESSION (276531@main): Content flash when expanding image viewer on eBay page
<a href="https://bugs.webkit.org/show_bug.cgi?id=275739">https://bugs.webkit.org/show_bug.cgi?id=275739</a>
<a href="https://rdar.apple.com/128178476">rdar://128178476</a>

Reviewed by Alan Baradlay.

Listings on eBay can display a photo gallery which is contained within a subtree where
the `hidden` presentation attribute is toggled to show and hide the gallery. The `hidden`
attribute really governs whether `display: none` is applied. In that subtree, a forwards-filling
`opacity` CSS Animation is applied.

Prior to `276531@main`, when `Style::TreeResolver::resolveElement()` would be called under
`Style::TreeResolver::resolveComposedTree()` for an element with `display: none`, an empty
`ElementUpdate` would be returned and `resolveComposedTree()` would not process its children.

In `276531@main`, we modified `resolveElement()` such that this behavior would not only apply
to elements with a newly-set `display: none` style associated with them, excluding those with
a cached `display: none` style. This was done in the static function `affectsRenderedSubtree()`.

This meant that now, if an element within a `display: none` subtree had its style modified,
`resolveElement()` would be called for that entire subtree when it previously wouldn&apos;t have.
As a result, `createAnimatedElementUpdate()` would now be called and since this method had
only ever been called for elements not contained within a `display: none` subtree, it would
not consider that possibility and start CSS Animations in that subtree.

This is what caused the eBay regression because the forwards-filling animation meant to fade
the photo gallery in would be reapplied soon after the photo gallery was hidden after an
additional style update, and when that photo gallery would become visible again, the animation
was already in effect and the gallery did not have the expected computed `opacity` value.

To fix this, we capture a new `isInDisplayNoneTree` state on the `Parent` objects created as
we iterate through the composed tree in `resolveComposedTree()`. That new member identifies
whether that element or a parent has `display: none` set on it. Now we can pass that state
down to `createAnimatedElementUpdate()` and ultimately to `Styleable::updateCSSAnimations()`
and ultimately neglect to start new CSS Animations when the element in question is contained
within a `display: none` subtree.

We also removed the `affectsRenderedSubtree()` changed introduced in `276531@main` which simply
did not make much sense, and to ensure that elements with `display: none` that had their style
updated correctly updated their cached `display: none` style, we now return a non-empty
`ElementUpdate` and add that `ElementUpdate` to `m_update` in `resolveComposedTree()`.

Finally, we add a new WPT test that reduces the eBay behavior.

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/display-none-prevents-starting-in-subtree-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/display-none-prevents-starting-in-subtree.html: Added.
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateElementRenderer):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::Parent::Parent):
(WebCore::Style::affectsRenderedSubtree):
(WebCore::Style::TreeResolver::resolveElement):
(WebCore::Style::TreeResolver::resolvePseudoElement):
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):
(WebCore::Style::TreeResolver::pushParent):
(WebCore::Style::TreeResolver::resolveComposedTree):
* Source/WebCore/style/StyleTreeResolver.h:
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::updateCSSAnimations const):
* Source/WebCore/style/Styleable.h:

Canonical link: <a href="https://commits.webkit.org/280876@main">https://commits.webkit.org/280876@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/60a152f5072a2c993dfb9c79545b61bbc38b8a3f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57801 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37129 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10277 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61423 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8246 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59929 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44765 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8434 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46831 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5854 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59831 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34897 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49953 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27660 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31593 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7246 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7250 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53543 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7516 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63106 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1715 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7588 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54053 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1721 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49964 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54176 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1458 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8636 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32958 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34044 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35128 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33789 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -266,6 +264,8 @@ auto TreeResolver::resolveElement(Element& element, const RenderStyle* existingS
 
     if (!affectsRenderedSubtree(element, *update.style)) {
+        if (update.style->display() == DisplayType::None && element.hasDisplayNone())
+            element.storeDisplayContentsOrNoneStyle(WTFMove(update.style));
```
This is not correct. TreeResolver must not store the style as it is an operation equivalent to render tree building. Instead it the style should be added to Style::Update and storing should be done by the storeDisplayContentsOrNoneStyle call in RenderTreeUpdater.

graouts committed:
REGRESSION (276531@main): Content flash when expanding image viewer on eBay page
https://bugs.webkit.org/show_bug.cgi?id=275739
rdar://128178476

Reviewed by Alan Baradlay.

Listings on eBay can display a photo gallery which is contained within a subtree where
the `hidden` presentation attribute is toggled to show and hide the gallery. The `hidden`
attribute really governs whether `display: none` is applied. In that subtree, a forwards-filling
`opacity` CSS Animation is applied.

Prior to `276531@main`, when `Style::TreeResolver::resolveElement()` would be called under
`Style::TreeResolver::resolveComposedTree()` for an element with `display: none`, an empty
`ElementUpdate` would be returned and `resolveComposedTree()` would not process its children.

In `276531@main`, we modified `resolveElement()` such that this behavior would not only apply
to elements with a newly-set `display: none` style associated with them, excluding those with
a cached `display: none` style. This was done in the static function `affectsRenderedSubtree()`.

This meant that now, if an element within a `display: none` subtree had its style modified,
`resolveElement()` would be called for that entire subtree when it previously wouldn't have.
As a result, `createAnimatedElementUpdate()` would now be called and since this method had
only ever been called for elements not contained within a `display: none` subtree, it would
not consider that possibility and start CSS Animations in that subtree.

This is what caused the eBay regression because the forwards-filling animation meant to fade
the photo gallery in would be reapplied soon after the photo gallery was hidden after an
additional style update, and when that photo gallery would become visible again, the animation
was already in effect and the gallery did not have the expected computed `opacity` value.

To fix this, we capture a new `isInDisplayNoneTree` state on the `Parent` objects created as
we iterate through the composed tree in `resolveComposedTree()`. That new member identifies
whether that element or a parent has `display: none` set on it. Now we can pass that state
down to `createAnimatedElementUpdate()` and ultimately to `Styleable::updateCSSAnimations()`
and ultimately neglect to start new CSS Animations when the element in question is contained
within a `display: none` subtree.

We also removed the `affectsRenderedSubtree()` changed introduced in `276531@main` which simply
did not make much sense, and to ensure that elements with `display: none` that had their style
updated correctly updated their cached `display: none` style, we now return a non-empty
`ElementUpdate` and add that `ElementUpdate` to `m_update` in `resolveComposedTree()`.

Finally, we add a new WPT test that reduces the eBay behavior.

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/display-none-prevents-starting-in-subtree-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/display-none-prevents-starting-in-subtree.html: Added.
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateElementRenderer):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::Parent::Parent):
(WebCore::Style::affectsRenderedSubtree):
(WebCore::Style::TreeResolver::resolveElement):
(WebCore::Style::TreeResolver::resolvePseudoElement):
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):
(WebCore::Style::TreeResolver::pushParent):
(WebCore::Style::TreeResolver::resolveComposedTree):
* Source/WebCore/style/StyleTreeResolver.h:
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::updateCSSAnimations const):
* Source/WebCore/style/Styleable.h:

Canonical link: https://commits.webkit.org/280876@main
